### ISSUE TITLE
Allow overriding the build method for Metrics

### ISSUE THREAD

**Issue by harahu** [tensorflow/tensorflow#40195]
**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): yes

**Describe the feature and the current behavior/state.**
Most `tf.keras.metrics.Metric` subclasses today are accumulated as scalars and hence can have their accumulation tensors statically initialized in `__init__` before the Metric is used. There are exceptions to this where tensor shape is dependent values only accessible during runtime. An example of an exception is the `MeanTensor` metric (https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/metrics.py#L2812-L2921), where the tensor shape is determined in a special `_build` method. That unfortunately uses protected parts of the API.

This was a problem for me when implementing/augmenting a custom RSquare metric for `tensorflow_addons` (https://www.tensorflow.org/addons/api_docs/python/tfa/metrics/RSquare). When performing multivariate regression, it is necessary to calculate the R2 metric for each variable separately, and hence I would really like to do something similar to what is done in `MeanTensor`, i.e. setting the shape of the accumulation tensors based on the de facto dimensionality of model outputs/labels during runtime. I shied away from this, however, due to the use of protected API components. The consequence is that the user of the Metric needs to specify the shape of the labels up front, which is error prone and annoying. Discussing this with one of the maintainers of `tensorflow_addons`, it was mentioned that there are reasons to expect `Metric` to properly support a `build` method in the future (https://github.com/tensorflow/addons/pull/1310#issuecomment-599143485). This feature request is partly to communicate this need, but with a focus on a specific use case rather than a specific implementation. I don't know if this has been covered in other issues. Perhaps you know, @gabrieldemarmiesse?

**Will this change the current api? How?**
I don't want to recommend any implementation over others as long as the use case is catered to: Dynamic shape metrics.

**Who will benefit with this feature?**
Developers and users of custom `tf.keras.metrics.Metric` subclasses.

**Any Other info.**

**Comment by gabrieldemarmiesse** [comment]
I did not encounter other issues, but since Metric is a subclass of layer, and layer has a "build" method, it makes sense to expose it since the use case is exactly the same for metrics and layers. It would lead to a cleaner API for metrics in general.

**Comment by gabrieldemarmiesse** [comment]
@harahu  the description of the issue is really good and I believe it's worth mentioning it in the next keras meeting, but the title of the issue can be misleading. Maybe "Allow overriding the build method for Metrics" would be accurate?

**Comment by harahu** [comment]
> @harahu the description of the issue is really good and I believe it's worth mentioning it in the next keras meeting, but the title of the issue can be misleading. Maybe "Allow overriding the build method for Metrics" would be more accurate?

Was trying to focus more on the use case rather than the implementation, but considering this is the most likely approach to solving this, your suggestion works as a title. Done. I really appreciate if this is brought up for discussion.

**Comment by gabrieldemarmiesse** [comment]
This was discussed in the last keras meeting, and we also talked about allowing `dynamic=True` for metrics. @fchollet agrees that both are good ideas and should be implemented. We're just lacking someone to implement it at this point. But a least it's just a question of time before we get `build` and `dynamic` in metrics :)

**Comment by harahu** [comment]
Glad to hear that! Let me know if I can be of assistance.

**Comment by mohantym** [comment]
Hi @harahu !

We are checking to see whether you still need help in this issue . Have you checked  [custom metrics](https://keras.io/api/metrics/#creating-custom-metrics) documentation with 2.8/2.10 version yet?

Thank you!

**Comment by harahu** [comment]
@mohantym There's nothing in the documentation you linked that indicates this issue having been addressed.

**Comment by mohantym** [comment]
@harahu !
Custom metrics documentation includes documentation for both stateless and stateful ( subclass of tf.keras.metrics.Metric) implementation.

You can create your own metrics object from tf.keras.metrics.Metric with [stateful implementation ](https://keras.io/api/metrics/#creating-custom-metrics:~:text=Here%27s%20a%20simple%20example%20computing%20binary%20true%20positives%3A)and call it during fit operation.


Thank you!

**Comment by harahu** [comment]
I could also create subclasses of `Metric` at the time this issue was created. That isn't what I'm asking for, and it is almost a prerequisite for what I _am_ asking for. I'm asking for tooling that allows the shape of a non-scalar metric to be computed at runtime in a sensible way, without having to ask the programmer to state this shape up front as an argument to `__init__`.

Currently, there's at least two patterns in use for doing this.
- `MeanTensor` [calls a custom implemented build method](https://github.com/tensorflow/tensorflow/blob/r2.10/tensorflow/python/keras/metrics.py#L3121-L3122) on the first call to `update_state`. This custom build method [relies on TensorFlow's protected API](https://github.com/tensorflow/tensorflow/blob/r2.10/tensorflow/python/keras/metrics.py#L3099), which implies it is not an example to be followed by other, external subclasses of `Metrics`.
- You [can check for missing attributes](https://github.com/tensorflow/addons/blob/r0.18/tensorflow_addons/metrics/r_square.py#L113-L140) during state update, like is done in the `RSquare` metric I originally requested this feature for. This pattern isn't documented anywhere I can find as a robust and safe choice.

What I am asking for, in abstract, is a clean, well-documented, canonical alternative approach to these two patterns. @gabrieldemarmiesse and @fchollet seem to have been discussing a `build` method for Metrics as a solution for this, as mentioned earlier in the thread.

Is what I'm asking clear to you now, or should I try to clarify further @mohantym?

**Comment by mohantym** [comment]
Ok @harahu !
Thanks for the update.  
I think this feature should be addressed in  [keras-team/keras repo](https://github.com/keras-team/keras/issues) then.
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999) . 
Thank you!

**Comment by harahu** [comment]
@mohantym I agree that this issue probably belongs in the keras repo. However, it was created before the split was made. I would have expected the organization to have made an effort in porting over any keras-related issues, including this one, if you feel community-created issues are valuable.

I take it this was not done. Would you like this issue to be ported over now? How do you suggest we go about that?

**Comment by mohantym** [comment]
No worries @harahu ! 
I was just suggesting to create another thread in Keras repo for better visibility . Just trying to contribute as much as possible.
Thanks for understanding.

**Comment by harahu** [comment]
> No worries @harahu ! I was just suggesting to create another thread in Keras repo for better visibility.

Good suggestion @mohantym. You are more than welcome to do that. I'd appreciate it. Let me know if you need any assistance.

**Comment by harahu** [comment]
@mohantym https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
